edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/BridgeConnection/MqttBridgeConnector.cs

```diff
@@ -0,0 +1,444 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Channels;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using uPLibrary.Networking.M2Mqtt;
+    using uPLibrary.Networking.M2Mqtt.Messages;
+
+    public class MqttBridgeConnector : IMqttBridgeConnector
+    {
+        const int ReconnectDelayMs = 2000;
+
+        readonly IComponentDiscovery components;
+        readonly ISubscriptionChangeHandler subscriptionChangeHandler;
+        readonly ISystemComponentIdProvider systemComponentIdProvider;
+        readonly Dictionary<ushort, TaskCompletionSource<bool>> pendingAcks = new Dictionary<ushort, TaskCompletionSource<bool>>();
+
+        readonly object guard = new object();
+
+        Option<Channel<MqttPublishInfo>> publications;
+        Option<Task> forwardingLoop;
+        Option<MqttClient> mqttClient;
+
+        public MqttBridgeConnector(IComponentDiscovery components, ISubscriptionChangeHandler subscriptionChangeHandler, ISystemComponentIdProvider systemComponentIdProvider)
+        {
+            this.components = components;
+            this.subscriptionChangeHandler = subscriptionChangeHandler;
+            this.systemComponentIdProvider = systemComponentIdProvider;
+
+            // because of the circular dependency between MqttBridgeConnector and the producers,
+            // in this loop the producers get the IMqttBridgeConnector reference:
+            foreach (var producer in components.Producers)
+            {
+                producer.SetConnector(this);
+            }
+        }
+
+        public async Task ConnectAsync(string serverAddress, int port)
+        {
+            Events.Starting();
+
+            var client = default(MqttClient);
+
+            lock (this.guard)
+            {
+                if (this.mqttClient.HasValue)
+                {
+                    Events.StartedWhenAlreadyRunning();
+                    throw new InvalidOperationException("Cannot start mqtt-bridge connector twice");
+                }
+                else
+                {
+                    this.mqttClient = Option.Some(new MqttClient(serverAddress, port, false, MqttSslProtocols.None, null, null));
+                }
+
+                client = this.mqttClient.Expect(() => new Exception("No mqtt-bridge connector instance found to setup"));
```

Comment:
A better pattern might be - 
client = new MqttClient()
this.mqttClient = Option.Some(client)